spirv_msl.cpp

```diff
@@ -854,65 +1047,370 @@ void CompilerMSL::emit_instruction(const Instruction &instruction)
-				emit_op(result_type, id,
-				        join("uint3(", img_exp, ".get_width(), ", img_exp, ".get_height(), ", img_exp, ".get_depth())"),
-				        false);
-				break;
+		if (length)
+		{
+			flags = *opt++;
+			length--;
+		}
 
-			default:
-				break;
+		auto test = [&](uint32_t &v, uint32_t flag) {
+			if (length && (flags & flag))
+			{
+				v = *opt++;
+				length--;
 			}
-		}
-		else
+		};
+
+		test(bias, ImageOperandsBiasMask);
+		test(lod, ImageOperandsLodMask);
+
+		statement(join(
+		    to_expression(img_id), ".write(", to_expression(texel_id), ", ",
+		    to_function_args(img_id, img_type, true, false, false, coord_id, 0, 0, 0, 0, lod, 0, 0, 0, 0, 0, &forward),
+		    ");"));
+
+		if (var && variable_storage_is_aliased(*var))
+			flush_all_aliased_variables();
+
+		break;
+	}
+
+	case OpImageQuerySize:
+	case OpImageQuerySizeLod:
+	{
+		uint32_t rslt_type_id = ops[0];
+		auto &rslt_type = get<SPIRType>(rslt_type_id);
+
+		uint32_t id = ops[1];
+
+		uint32_t img_id = ops[2];
+		string img_exp = to_expression(img_id);
+		auto &img_type = expression_type(img_id);
+		Dim img_dim = img_type.image.dim;
+		bool is_array = img_type.image.arrayed;
+
+		if (img_type.basetype != SPIRType::Image)
 			SPIRV_CROSS_THROW("Invalid type for OpImageQuerySize.");
+
+		string lod;
+		if (opcode == OpImageQuerySizeLod)
+		{
+			// LOD index defaults to zero, so don't bother outputing level zero index
+			string decl_lod = to_expression(ops[3]);
+			if (decl_lod != "0")
+				lod = decl_lod;
+		}
+
+		string expr = type_to_glsl(rslt_type) + "(";
+		expr += img_exp + ".get_width(" + lod + ")";
+
+		if (img_dim == Dim2D || img_dim == DimCube || img_dim == Dim3D)
+			expr += ", " + img_exp + ".get_height(" + lod + ")";
+
+		if (img_dim == Dim3D)
+			expr += ", " + img_exp + ".get_depth(" + lod + ")";
+
+		if (is_array)
+			expr += ", " + img_exp + ".get_array_size()";
+
+		expr += ")";
+
+		emit_op(rslt_type_id, id, expr, should_forward(img_id));
+
 		break;
 	}
 
+#define ImgQry(qrytype)                                                                     \
+	{                                                                                       \
+		uint32_t rslt_type_id = ops[0];                                                     \
+		auto &rslt_type = get<SPIRType>(rslt_type_id);                                      \
+		uint32_t id = ops[1];                                                               \
+		uint32_t img_id = ops[2];                                                           \
+		string img_exp = to_expression(img_id);                                             \
+		string expr = type_to_glsl(rslt_type) + "(" + img_exp + ".get_num_" #qrytype "())"; \
+		emit_op(rslt_type_id, id, expr, should_forward(img_id));                            \
+		break;                                                                              \
+	}
+
+	case OpImageQueryLevels:
+		ImgQry(mip_levels)
+
+		    case OpImageQuerySamples : ImgQry(samples)
```

Comment:
You'll probably want to wrap this block in a do/while here so you can do `ImgQry(samples);` with a semi-colon to not confuse clang-format.